src/ott/neural/models/base_solver.py

```diff
@@ -0,0 +1,485 @@
+      self,
+      tau_a: float,
+      tau_b: float,
+      resample_epsilon: float,
+      scale_cost: Union[bool, int, float, Literal["mean", "max_cost",
+                                                  "median"]] = "mean",
+      sinkhorn_kwargs: Dict[str, Any] = types.MappingProxyType({}),
+  ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    """Compute the unbalanced source and target marginals for a batch."""
+
+    @jax.jit
+    def compute_unbalanced_marginals(
+        batch_source: jnp.ndarray, batch_target: jnp.ndarray
+    ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+      geom = pointcloud.PointCloud(
+          batch_source,
+          batch_target,
+          epsilon=resample_epsilon,
+          scale_cost=scale_cost
+      )
+      out = sinkhorn.Sinkhorn(**sinkhorn_kwargs)(
+          linear_problem.LinearProblem(geom, tau_a=tau_a, tau_b=tau_b)
+      )
+      return out.matrix.sum(axis=1), out.matrix.sum(axis=0)
+
+    return compute_unbalanced_marginals
+
+  @jax.jit
+  def _resample_unbalanced(
+      self,
+      key: jax.random.KeyArray,
+      batch: Tuple[jnp.ndarray, ...],
+      marginals: jnp.ndarray,
+  ) -> Tuple[jnp.ndarray, ...]:
+    """Resample a batch based on marginals."""
+    indices = jax.random.choice(
+        key, a=len(marginals), p=jnp.squeeze(marginals), shape=[len(marginals)]
+    )
+    return tuple(b[indices] if b is not None else None for b in batch)
+
+  def _setup(self, source_dim: int, target_dim: int, cond_dim: int):
+    self.rng_unbalanced, rng_eta, rng_xi = jax.random.split(
+        self.rng_unbalanced, 3
+    )
+    self.unbalancedness_step_fn = self._get_rescaling_step_fn()
+    if self.rescaling_a is not None:
+      self.opt_eta = (
+          self.opt_eta if self.opt_eta is not None else
+          optax.adamw(learning_rate=1e-4, weight_decay=1e-10)
+      )
+      self.state_eta = self.rescaling_a.create_train_state(
+          rng_eta, self.opt_eta, source_dim
+      )
+    if self.rescaling_b is not None:
+      self.opt_xi = (
+          self.opt_xi if self.opt_xi is not None else
+          optax.adamw(learning_rate=1e-4, weight_decay=1e-10)
+      )
+      self.state_xi = self.rescaling_b.create_train_state(
+          rng_xi, self.opt_xi, target_dim
+      )
+
+  def _get_rescaling_step_fn(self) -> Callable:  # type:ignore[type-arg]
+
+    def loss_a_fn(
+        params_eta: Optional[jnp.ndarray],
+        apply_fn_eta: Callable[[Dict[str, jnp.ndarray], jnp.ndarray],
+                               jnp.ndarray],
+        x: jnp.ndarray,
+        condition: Optional[jnp.ndarray],
+        a: jnp.ndarray,
+        expectation_reweighting: float,
+    ) -> Tuple[float, jnp.ndarray]:
+      eta_predictions = apply_fn_eta({"params": params_eta}, x, condition)
+      return (
+          optax.l2_loss(eta_predictions[:, 0], a).mean() +
+          optax.l2_loss(jnp.mean(eta_predictions) - expectation_reweighting),
+          eta_predictions,
+      )
+
+    def loss_b_fn(
```

Comment:
This function is almost the same as `loss_a_fn`, consider refactoring them into 1 function.